Reorganize template dropdown menu and refactor `TemplateOptionsController`

momo3404 opened the pull request:
Fixes #1196  .

Changes proposed in this PR:
- Change the ordering of templates in the template dropdown menu from alphabetical to all organizational templates first, followed by the  Alliance Simplified Template, Alliance Template, and all remaining templates.
- Simplified template loading logic by removing explicit `Template.default`  handling.
- `Template.latest_customizable.where(org_id: funder.id)` and `Template.published.publicly_visible.where(org_id: funder.id, customization_of: nil)` now fetch the default template as well
- Removed manual insertion and filtering of the default template; `sort_templates()` now determines ordering.

momo3404 committed:
Order templates in dropdown properly

momo3404 committed:
Move sorting code to separate function

aaronskiba commented:
The ordering seems to be off in the dropdown when I select `Digital Research Alliance of Canada` as the Organisation.

aaronskiba commented on app/controllers/template_options_controller.rb:
```diff
@@ -84,4 +85,22 @@ def plan_params
+  # Assign the order of templates in the dropdown menu
+  def sort_templates(templates, org) # rubocop:disable Metrics/AbcSize
+    # Get all templates belonging to the selected organization
+    org_templates = templates.select { |template| template.org_id == org&.id && template.customization_of.nil? }
```
Why are you using `template.customization_of.nil?`?

momo3404 replied:
It seemed to fix this bug that occurs with some orgs (like McMaster for example) where the order would be Alliance Simplified Template -> Org templates -> Alliance Template

aaronskiba replied:
I think this query result might reveal a bit:
<img width="470" height="290" alt="Screenshot from 2025-10-02 12-13-15" src="https://github.com/user-attachments/assets/1173e486-48ff-4b25-adc4-0ebbe480a540" />

Non-Alliance orgs may have their own customisation of the "Alliance Simplified Template". If an org does have that customisation, then it will be included in `org_templates`. And further, I'm thinking the `.uniq` execution at the end would remove the `alliance_template = templates.find { |t| t.title == 'Alliance Template' }` duplicate.

aaronskiba replied:
@CeriMacMillan, what to do in the following scenario?

An org has customised the `Alliance Simplified Template (Funding Application Stage)` template but the title remains identical. Should this be grouped with the org templates? Or should it be grouped below that in the (Alliance Simplified Template (Funding Application Stage) / Alliance Template) section?

Maybe that same question can be asked of any org customisation of an Alliance template the retains the exact same title.

CeriMacMillan replied:
@aaronskiba That's a really good question, I would lean towards putting it with the org templates as they have taken the time to add local content/context. Maybe they can come after any or templates named as 'org template' but before the Alliance templates. 

Can double check this with @closenma though to be sure.

closenma replied:
I agree, if it has customization then let's count it at organizational and sort it above the other Alliance templates.

momo3404 commented:
> The ordering seems to be off in the dropdown when I select `Digital Research Alliance of Canada` as the Organisation.

I'm guessing that's because the Alliance templates get included in `org_templates`. Just to confirm, the order you're seeing is Alliance Template -> Two other Alliance templates -> Alliance Simplified Template?

aaronskiba commented:
> > The ordering seems to be off in the dropdown when I select `Digital Research Alliance of Canada` as the Organisation.
> 
> I'm guessing that's because the Alliance templates get included in `org_templates`. Just to confirm, the order you're seeing is Alliance Template -> Two other Alliance templates -> Alliance Simplified Template?

That's right. Basically, Alliance Template, and then all of the others in alphabetical order.

momo3404 committed:
Fix alliance simplified template issue and alliance org selection

- With these changes, if the chosen org is the Alliance, now the first two displayed templates are the Alliance Simplified and Alliance main templates
- The previous code incorrectly excluded org templates that were customizations of other templates. Now those templates are included at the top of the dropdown menu along with other org templates.

aaronskiba commented on app/controllers/template_options_controller.rb:
```diff
@@ -87,20 +87,30 @@ def org_params
+  # The desired order of templates in the dropdown is:
+  # All organizational templates first followed by the Alliance Simplified Template
+  # Followed by the Alliance Template and then all remaining templates
+  def sort_templates(templates, org) # rubocop:disable Metrics/AbcSize,Metrics/PerceivedComplexity
```
Maybe we could refactor this a bit more to address the rubocop offences? Maybe something like the following added method would help?

```ruby
def alliance_templates(templates)
  simplified = templates.find { |t| t.title.start_with?('Alliance Simplified Template') }
  alliance   = templates.find { |t| t.title == 'Alliance Template' }
  [simplified, alliance].compact
end
```

momo3404 committed:
Add `alliance_templates` function and fix rubocop offenses

aaronskiba reviewed (commented):
This snippet mutates `@templates` by removing default/customization and then prepending it:

```ruby
      @templates.select! { |t| t.id != Template.default.id && t.id != customization.id }
      # We want the default template to appear at the beggining of the list
      @templates.unshift(customization)
    end
```

That snippet was used for the prior ordering of the templates. Since `sort_templates` is now fully responsible for the ordering, maybe we should address/remove the old ordering code.

momo3404 commented:
> This snippet mutates `@templates` by removing default/customization and then prepending it:
> 
> ```ruby
>       @templates.select! { |t| t.id != Template.default.id && t.id != customization.id }
>       # We want the default template to appear at the beggining of the list
>       @templates.unshift(customization)
>     end
> ```
> 
> That snippet was used for the prior ordering of the templates. Since `sort_templates` is now fully responsible for the ordering, maybe we should address/remove the old ordering code.

I think that code might still be necessary. I tried removing it, but with that change the Alliance Template now never shows up in the dropdown. 

I also am not sure how much we should meddle with the existing code, since it seems there's a lot going on with customizations and default templates.

aaronskiba commented:
> > This snippet mutates `@templates` by removing default/customization and then prepending it:
> > ```ruby
> >       @templates.select! { |t| t.id != Template.default.id && t.id != customization.id }
> >       # We want the default template to appear at the beggining of the list
> >       @templates.unshift(customization)
> >     end
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That snippet was used for the prior ordering of the templates. Since `sort_templates` is now fully responsible for the ordering, maybe we should address/remove the old ordering code.
> 
> I think that code might still be necessary. I tried removing it, but with that change the Alliance Template now never shows up in the dropdown.
> 
> I also am not sure how much we should meddle with the existing code, since it seems there's a lot going on with customizations and default templates.

Yes, this controller is looking pretty fragile. I might take a shot at that change though, just to help a bit with future maintainability.

aaronskiba committed:
Clean up comments in TemplateOptionsController

aaronskiba committed:
Refactor: Include default template in funder templates query

Simplified template loading logic by removing explicit default template handling.
- `Template.latest_customizable.where(org_id: funder.id)` and `Template.published.publicly_visible.where(org_id: funder.id, customization_of: nil)` now fetch the default template as well
- Removed manual insertion and filtering of the default template; `sort_templates()` now determines ordering.

aaronskiba committed:
Fix test: Associate default template with default funder

The changes in a4fc95edf50f91f7647e3891948d28e7c2d9fc43 require that the default template belong to the default org.

momo3404 commented on app/controllers/template_options_controller.rb:
```diff
@@ -84,4 +85,36 @@ def plan_params
   end
+
+  def alliance_templates(templates)
+    simplified = templates.find { |t| t.title.start_with?('Alliance Simplified Template') }
```
I wonder if we should add a check here to ensure that the simplified template belongs to the Alliance since there seems to be many templates with the same title.

aaronskiba replied:
Yes, that's a good idea. Similarly, I'd recommend using `Rails.application.config.default_funder_id` to verify the funder org rather than `org&.name == 'Digital Research Alliance of Canada'`. It also might be better to use the `is_default` check when fetching the non-simplified "Alliance Template".

momo3404 committed:
Add org check for alliance templates and use funder_id

aaronskiba committed:
Refactor `sort_templates()`

aaronskiba reviewed (dismissed):
Looks good. Sorting looks in order and I'm liking the simpler handling of the default template throughout this controller now.

momo3404 committed:
Add changelog entry
